[ADD] sale_coupon_multi_use: init

oerp-odoo opened the pull request:
Allows to consume same coupon multiple times. Can only be used when
discount has fixed amount. Discount fixed amount can't be changed if
there are coupons with `multi_use` flag set.

Can specify value of coupon and that value can be consumed per multiple
sale orders.

If discount is removed from sale order, so is consumption line from
coupon.

oerp-odoo committed:
[ADD] sale_coupon_multi_use: init

Allows to consume same coupon multiple times. Can only be used when
discount has fixed amount. Discount fixed amount can't be changed if
there are coupons with multi_use flag set.

Can specify value of coupon and that value can be consumed per multiple
sale orders.

If discount is removed from sale order, so is consumption line from
coupon.

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+
+    # Takes value from related program (coupon_multi_use field), when
+    # it is generated.
+    multi_use = fields.Boolean("Multi Use", readonly=True)
```
```suggestion
    multi_use = fields.Boolean(readonly=True)
```
fields string would be composed automatically like that

oerp-odoo replied:
I know, just like to be explicit here:)

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+        return super().write(vals)
+
+
+class SaleCouponConsumptionLine(models.Model):
```
Please put one model per file.

oerp-odoo replied:
Hm, I was using this guidelines: https://www.odoo.com/documentation/13.0/reference/guidelines.html#file-naming
It allows to group models, following main one, (like putting sale.order.line on sale_order.py, if there is also sale.order used - which is main model).

You explicitly split every model?

rousseldenis replied:
@oerp-odoo Please read OCA conventions https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#15file-naming

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+    sale_order_line_id = fields.Many2one(
+        "sale.order.line", "Sale Order Line", required=True, ondelete="cascade"
+    )
+    amount = fields.Float()
```
Please use digits precision

oerp-odoo replied:
Thing is, I did this because float fields on sale_coupon, also do not use precision. So used same pattern here. Now then if I would use precision, I should probably use configurable decimal accuracy. But then its not clear which one would fir here, Price?

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,61 @@
+
+    def _get_multi_use_coupons(self):
+        self.ensure_one()
+        return self.coupon_ids.filtered(lambda r: r.multi_use)
```
```suggestion
        return self.coupon_ids.filtered("multi_use")
```

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+        sol = self.sale_order_line_id
+        sol_discount = abs(sol.price_unit)
+        amount_to_adjust = sol_discount - self.amount
+        if amount_to_adjust > 0:
```
Please use float_compare()

oerp-odoo replied:
And precision to use? The one from related product UOM?

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+    consumption_line_ids = fields.One2many(
+        "sale.coupon.consumption_line", "coupon_id", "Consumption Lines", readonly=True,
+    )
+    discount_fixed_amount_delta = fields.Float(
```
Please use digits precision.

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+            len(self) == 1
+            and self.multi_use
+            # Must have amount to split.
+            and self.discount_fixed_amount_delta > 0
```
Please use float_compare()

rousseldenis commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+        new_delta, discount = self._get_compared_discount_with_delta(coupon_order_data)
+        self._consume_line(discount, coupon_order_data["order"])
+        # Indicates whether coupon was fully consumed.
+        return new_delta <= 0
```
Please use float_compare()

rousseldenis commented on sale_coupon_multi_use/wizards/sale_coupon_apply_code.py:
```diff
@@ -0,0 +1,19 @@
+    def apply_coupon(self, order, coupon_code):
+        """Extend to pass order coupon ctx for multi coupon usage."""
+        self = self.with_context(
+            coupon_order_data={
```
I don't really like using context to pass arguments. It was not done for that purpose and should be avoided as much as possible (I know the original apply_coupon is not easily overridable)

oerp-odoo replied:
So any suggestion, to extend it better? 
Because in my opinion otherwise you would need to overwrite copy/paste blocks of code, which is ugly and not great.
Also, for what else, context is meant, if not passing stuff around? It does just pass stuff:) Though I agree its not as obvious as if you would pass proper argument from method.

I use this rule of thumb (priority):
- Extend method normally and pass/modify it without context (like for example, you can update result dictionary or something).
- If can't reach needed insides of a method by extending normally, pass context to intercept wanted parts.
- if context can't be used (not reachable, poor original implementation etc), then use last resort override whole method.

rousseldenis replied:
No, I've not digged into code too much. If you estimate this is the only way to do it, let's go. Just change the other things.

rousseldenis reviewed (changes requested):
@oerp-odoo Thanks for your contribution.

Code review

i-vyshnevska commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+
+    def _get_discount_fixed_amount_delta(self):
+        self.ensure_one()
+        amount_total = self.program_id.discount_fixed_amount
```
for additional saftines I would make "discount_fixed_amount" readonly if program has one used coupon

oerp-odoo replied:
There is constraint to not allow changing it, if there is at least one. 

Though the way Odoo implemented program and coupons relations is quite clunky. I think better way would be, if program would act just as a template and then coupons would copy values. So it would be independent from later program changes. But that would require some more improving.

For example in standard sale_coupon, when you apply some coupon discount on SO, it creates SOL with product on program. If later you change discount amount on program, it updates related product name, mentioning different amount. And then when you open that old SO again, it will show that new amount on product name as if that discount was applied, when actually it was not. So its kind of confusing to user I think.

i-vyshnevska commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -0,0 +1,154 @@
+                coupon_consumed = self._handle_multi_use(coupon_order_data)
+                if not coupon_consumed:
+                    # Makes it so coupon that coupon is not used up yet.
+                    del vals["state"]
```
do we need to add a new state option to state field? smth kind of "started"

oerp-odoo replied:
That could work, but then we would introduce additional complexity, because sale_coupon does not expect new state. That would mean, we would need to make sure it would understand what that state means (might be multiple places where we would need to intercept).

oerp-odoo committed:
[FIX] sale_coupon_multi_use: sale confirm (#20)

Coupons are marked as consumed twice. First time when coupon is added
and second when sale order is confirmed.

For that, multi-use coupon must not be marked as consumed when
confirming order (if it is still valid by multi-use rules).

Also had to redesign the way multi use coupons are checked on write,
because during confirmation it can update multiple coupons and old
implementation was able to handle one coupon at a time.

On top of that, simplified multi-use coupon handling. Now correct amount
is intercepted when discount_fixed_amount is computed for reward line.

This gives multiple advantages:
    - no need to adjust sale order line discount afterwards.
    - _compute_program_amount handles currency conversion, so multi-use
        consumed amount now also takes care of that too.
    - simpler consumption line creation and checking if coupon is fully
        consumed (no need to compare delta with discount that would need
        to be computed from sale order).

[FIX] sale_coupon_multi_use: action_cancel

Make sure related consumption lines are removed if coupon is being
reset (made valid again), e.g. SO is being cancelled.

[REF] sale_coupon_multi_use: LIC/naming/misc/lang

oerp-odoo committed:
Sale Coupon Multi-Currency  (#21)

* [ADD] sale_coupon_multi_currency: init

Makes currency field visible on coupon/promotion program.

Can set custom currency instead of just using company currency.

* [IMP] sale_coupon_multi_use: multi-currency support

Multi Use coupons are now multi currency aware and will convert used
amounts back to program currency.

* [REF] sale_coupon_multi|use|currency

Decoupled dependency between sale_coupon_multi_use and
sale_coupon_multi_currency, so it could be used independently, but also
added bridge module to make sure currency can't be changed if multi
use coupon is already used.

Also fixed currency conversion bug from coupon to sale order amount -
conversion must have been done before comparing SO amount not after,
because amounts compared could be in different currencies.

oerp-odoo committed:
[IMP] sale_coupon_multi_use: m2m relation (#23)

`coupon_multi_use_ids` on sale.order is now used to track multi-use
coupons on sale orders. `applied_coupon_ids` field did not suit multi-use
coupons case, because it is o2m field where if same coupon is applied
on next sale order, its relation with old sale order is removed.

This relation follows similar logis as is for `applied_coupon_ids` field
relation logic.

* [FIX] sale_coupon_multi_use: action_cancel

Odoo now fixed the way coupons are treated once sale order is cancelled.
Now when SO is cancelled, coupons are unattached from SO. In this case
action_draft support for multi-use is no longer needed, because when
SO is cancelled, same coupon should be readded manually.

jcoux committed:
sale_coupon_multi_use: Rewrite module (#31)

jcoux commented:
Changed since last reviews:
* Totally rewrite the module `sale_coupon_multi_use` to make it more easy to read and maintainable.
* Fix bugs on module `sale_coupon_multi_use`.
* Add module `sale_coupon_multi_currency`.
* Add module `sale_coupon_multi_use_currency` to have a glue between 2 other modules.

damdam-s commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -0,0 +1,164 @@
+#: code:addons/sale_coupon_multi_use/models/sale_coupon_program.py:0
+#, python-format
+msgid "Coupon multi use can't be changed with existing coupons."
+"Le bon de réduction sécable ne peut pas être modifié avec des bons de réduction existants."
```
```suggestion
msgstr "Le bon de réduction sécable ne peut pas être modifié avec des bons de réduction existants."
```

damdam-s reviewed (approved):
LG

epanisset reviewed (commented):
Functional test done: Ok approved.

jcoux commented:
> Changed since last reviews:
> 
> * Totally rewrite the module `sale_coupon_multi_use` to make it more easy to read and maintainable.
> * Fix bugs on module `sale_coupon_multi_use`.
> * Add module `sale_coupon_multi_currency`.
> * Add module `sale_coupon_multi_use_currency` to have a glue between 2 other modules.

@rousseldenis As said in my previous comment: a lot of changes on this PR.
And I now have 2 technical approvals + 1 functional approval.
Can you update (or maybe view the changes redo) a review to allow this PR can be merged?

Thanks in advance! :)

jcoux commented:
Hello @rousseldenis
If you could update your review, it would can very cool: I would like to merge this PR!
Thanks in advance.

rousseldenis reviewed (approved):
Code review

rousseldenis commented:
@jcoux Are these one ready ?

jcoux commented:
@rousseldenis 
Thanks a lot for reviews! :)
For me yes, it's ready.
Just asked me if it's better to squash all commits or not.
What is your opinion about this?

rousseldenis commented:
> @rousseldenis
> Thanks a lot for reviews! :)
> For me yes, it's ready.
> Just asked me if it's better to squash all commits or not.
> What is your opinion about this?

My opinion is that commits should reflect development flow. So, if they make sense, keep them, if some can be squashed together, do it. Some people don't like too many commits, I'm not if they means something.

jcoux commented:
As this module is relatively complex and I rewritten a big part of it, I think it's better to keep commit history like we have now: for me we can merge ;)

rousseldenis commented:
/ocabot merge patch
